docs/source/guides/schema-design.md

```diff
@@ -277,9 +258,86 @@ Let’s break this down, field by field:
+
+```graphql
+type LikePostMutationResponse implements MutationResponse {
+  code: String!
+  success: Boolean!
+  message: String!
+  post: Post
+  user: User
+}
+```
+
+In this response type, we've provided the expectation that both the `user` and the `post` would be returned and an actual response to a `likePost` mutation could be:
+
+```json
+{
+  "data": {
+    "likePost": {
+      "code": "200",
+      "success": true,
+      "message": "Thanks!",
+      "post": {
+        "likes": 5040
+      },
+      "user": {
+        "reputation": 11
+      }
+    }
+  }
+}
+```
+
+Following this pattern for mutations provides detailed information about the data that has changed and feedback on whether the operation was successful or not.  Armed with this information, developers can easily react to failures within the client
+
+<h3 id="mutation-input-types">Input types</h3>
+
+Input types are a special type in GraphQL which allows an object to be passed as an argument to both queries and mutations and is helpful when simple scalar types aren't sufficient.
+
+This allows arguments to be structured in an more manageable way, similar to how switching to an `options` argument might be appreciated when `function` arguments become too iterative.
+
+For example, consider this mutation which creates a post along with its accompanying media URLs (e.g. images):
+
+```graphql
+type Mutation {
+  createPost(title: String, body: String, mediaUrls: [String]): Post
+}
+```
+
+This could be easier to digest, and the arguments would be easier to re-use within the mutation, by using an `input` type with the relevant fields.
+
+An input type is defined like a normal object type but using the `input` keyword.  To introduce an `input` type for this example, we'd do:
+
+```graphql
+type Mutation {
+  createPost(post: PostAndMediaInput): Post
+}
+
+input PostAndMediaInput {
+  title: String
+  body: String
+  mediaUrls: [String]
+}
+```
+
+Not only does this facilitate passing the `PostAndMediaInput` around within the schema, it also provides a basis for annotating fields with descriptions which are automatically exposed by GraphQL-enabled tools:
+
+```graphql
+input PostAndMediaInput {
+  "A main title for the post"
+  title: String
+  "The textual body of the post."
+  body: String
+  "A list of URLs to render in the post."
+  mediaUrls: [String]
+}
+```
 
-Following this pattern for mutations provides detailed information about the data that has changed and feedback on whether the operation was successful or not.  Armed with this information, developers can easily react to failures in the client and fetch the information they need to update their local cache.
+Input types can also be used when different operations require the exact same information, though we urge caution on over-using this technique since changes to `input` types are breaking changes for all operations which utilize them.
```

Comment:
This looks great! Maybe we could add a sentence here cautioning readers about reusing input types for queries and mutations since it's likely that some fields would be nullable on one but not the other?